[FEATURE] fast ostream buf iterator

smehringer opened the pull request:
Analoguesly to the fast_istreambuf_iterator, this ostreambuf iterator does less virtual function calls.

Still WIP!

Resolves #1570

smehringer commented:
@h-2 
If you have time, do you want to take a look and maybe repeat your benchmark? 
This may improve performance. (If I did not screw up, ... I have no clue if I understood it correctly)

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -135,7 +136,8 @@ class format_fasta
     {
-        seqan3::ostreambuf_iterator stream_it{stream};
+        auto stream_view = seqan3::views::ostreambuf(stream);
+        auto stream_it = stream_view.begin();
```
I don't think we need `views::ostreambuf` at all since there are no other "pure output views" currently and we never compare with end.

Just use the iterator directly like before!

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +166,119 @@ class fast_istreambuf_iterator
+        return eof;
+    }
+
+    /*!\name Comparison operators
```
pure output iterators don't need these at all.

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -135,7 +135,7 @@ class format_fasta
                                id_type && id,
```
Note that #1453 (waiting for review by you :wink) also touches these files!

smehringer replied:
My touch is way lighter :smirk_cat: Can you benchmark after my chagne again and see if we still need the extensive changes you introduce in #1453 ?

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,39 @@ class format_fasta
-            }
-            if (count % options.fasta_letters_per_line != 0)
+                stream_it.write_n(it, options.fasta_letters_per_line);
                 detail::write_eol(stream_it, options.add_carriage_return);
```
Are you not writing the first line of the sequence again and again here? Were do you increment `it`?
Is splitting the sequence and writing it line-wise better than chunk | join and writing as a whole? Could be either...

smehringer replied:
> Are you not writing the first line of the sequence again and again here? Were do you increment it?

`write_n` increments `it`.. I already thought that this might be a design flaw. I can also return the updated iterator instead?

> Is splitting the sequence and writing it line-wise better than chunk | join and writing as a whole? Could be either...

Well, your code explicitly stated that a for loop is faster than view chunk join and changing the ostreambuf iterator won't change that? So I would suggest to only keep the most efficient code even though the range version is more pretty.
I do not have time for extensive benchmarks on this :/

h-2 replied:
> write_n increments it.. I already thought that this might be a design flaw. I can also return the updated iterator instead?

Yes, that would be better! Iterators are usually passed by copy or move, not by reference.

> Well, your code explicitly stated that a for loop is faster than view chunk join and changing the ostreambuf iterator won't change that? So I would suggest to only keep the most efficient code even though the range version is more pretty. I do not have time for extensive benchmarks on this :/

Well, the point about my code was that you *can* easily benchmark it by turning a macro on/off :nerd_face: 

If I am to benchmark this again, it would really be a lot easier, if you only replace the lines that have `copy()` with `write_range()`...

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -19,6 +21,7 @@
 #endif // __cpp_lib_ranges
 
 #include <seqan3/core/platform.hpp>
+#include <seqan3/range/views/drop.hpp>
```
Where is this used?

smehringer replied:
It was used in your version of write_range. I forgot to remove it again.

smehringer replied:
it is used now

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+    using value_type        = char_t;                  //!< The char type of the stream.
+    using reference         = char_t;                  //!< The char type of the stream.
+    using pointer           = void;                    //!< Has no pointer type.
+    using iterator_category = std::input_iterator_tag; //!< Pure input iterator.
```
```suggestion
    using iterator_category = std::output_iterator_tag; //!< Pure output iterator.
```

smehringer replied:
ops

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+                throw std::ios_base::failure{"Could not write to output stream."};
+
+            ++it;
+            ++buffer_space;
```
why not just call overflow?

h-2 replied:
or `*this = *it;` ?

smehringer replied:
> why not just call overflow?

I do call overflow in the if clause above this? I just increment it here because overflow has written the character already (and I cannot call `overflow()` without a character because it won't actually overflow then). I could not increment buffer_space here but instead do `write_n(it, n - buffer_space + 1);` in the line below.

> or *this = *it; ?

Because this would do another if check within `operator=()` but I already know that I am at the end of my buffer :woman_shrugging:

h-2 replied:
> I cannot call overflow() without a character because it won't actually overflow then

Ah, ok :+1: 

I thought it was strange that the `== eof() and  throw`  had disappeared from the other code blocks but reappeared here now. But, up to you!

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+    {
+        auto it = rng.begin();
+        write_n(it, std::ranges::distance(rng));
+    }
```
Why do we need two functions (write_n and write_range)?

smehringer replied:
Convenience?
(`write_n` is more generic and good for chunked output like fasta-line-wrapping without using `view::take` all the time. But write_range is more convinient in general)

h-2 replied:
I think, I would have designed it as a single function since view_take is free for random access ranges and even single iterators can be turned into such a range via `views::counted`, but that's up to you!

smehringer replied:
Well but as you said earlier (that distance() might be expensive for non sized views) we do not always have random access ranges, so I thought this is more efficient ?

h-2 replied:
It's hard to say without benchmarking to be honest.

smehringer replied:
So I did a benchmark and the time was the same so I removed write_n now

h-2 replied:
:+1:

smehringer commented:
Lowlevel benchmarks

```
------------------------------------------------------------------------------
Benchmark                                    Time             CPU   Iterations
------------------------------------------------------------------------------
read_all<tag::std_stream_it>           7122309 ns      7121778 ns           97
read_all<tag::std_streambuf_it>        1522230 ns      1522175 ns          461
read_all<tag::seqan3_streambuf_it>      531766 ns       531724 ns         1291
read_all<tag::seqan2_stream_it>         378025 ns       378010 ns         1842
write_all<tag::std_stream_it>           191470 ns       180745 ns         3890
write_all<tag::std_streambuf_it>         58340 ns        45578 ns        15905
write_all<tag::seqan3_streambuf_it>      57148 ns        46310 ns        16013
write_all<tag::seqan2_stream_it>         74148 ns        45971 ns        15493

```

smehringer commented:
@h-2 Why is my iterator not an `std::ostream_iterator` :cry:

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,39 @@ class format_fasta
-            size_t count = 0;
-            for (auto c : seq)
+            auto it = v.begin();
+            distance_type num_lines = std::ranges::distance(seq) / options.fasta_letters_per_line;
```
Calling distance might be really expensive if the range is not sized! Fasta files can hold genomes...

smehringer replied:
Well but without knowing the size, I cannot efficiently use the buffer.. I could differentiate between both cases..

smehringer replied:
I removed the dependance on std::distance

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+        }
+        else
+        {
+            it = std::ranges::copy_n(it, buffer_space, stream_buf->pptr()).in;
```
```suggestion
            /* Fill up remaining buffer */
            it = std::ranges::copy_n(it, buffer_space, stream_buf->pptr()).in;
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+            it = std::ranges::copy_n(it, buffer_space, stream_buf->pptr()).in;
+            stream_buf->pbump(buffer_space);
+
+            if (stream_buf->overflow(*it) == traits_t::eof()) // update put area by trying to write the next char
```
```suggestion
            /* Push one more character and flush */
            if (stream_buf->overflow(*it) == traits_t::eof()) // update put area by trying to write the next char
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +169,138 @@ class fast_istreambuf_iterator
+            ++it;
+            ++buffer_space;
+
+            write_n(it, n - buffer_space);
```
```suggestion
            /* Write remaining input data recursively */
            write_n(it, n - buffer_space);
```

h-2 replied:
Some documentation would help understanding the semantics here!

h-2 replied:
You removed the documentation again in the latest version :roll_eyes:

smehringer replied:
well because I removed write_n?

h-2 commented:
> Lowlevel benchmarks

It's slower for single character writes than the std::ostreambuf_iterator in microbenchmarks? :thinking: 
Did you have a look at LLVM's oder GNU's ostreambuf implementation by chance?

> EDIT: with write range

That looks good!

smehringer commented:
> It's slower for single character writes than the std::ostreambuf_iterator in microbenchmarks? thinking
Did you have a look at LLVM's oder GNU's ostreambuf implementation by chance?

no it's not ? It's roughly equal

h-2 commented:
> no it's not ? It's roughly equal

It's even 5% faster, I don't know what confused me there :man_facepalming:

h-2 commented on include/seqan3/io/sequence_file/format_embl.hpp:
```diff
@@ -245,34 +255,51 @@ class format_embl
+            auto it = stream_it.write_range(char_sequence | views::take(10)); // write first 10 characters
+            stream_it = ' ';
+
+            while (it != char_sequence.end())
```
Do not rely on member functions when you are not sure about the type. We make no promises regarding what `views::to_char` returns so it's better to use `std::ranges::begin|end|...`.

h-2 replied:
This affects quite a few places!

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,35 @@ class format_fasta
+    template <typename stream_it_t, typename seq_type>
+    void write_seq(stream_it_t & stream_it, sequence_file_output_options const & options, seq_type && seq)
     {
+        // using distance_type = decltype(std::ranges::distance(seq));
```
```suggestion
```

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,35 @@ class format_fasta
-            detail::write_eol(stream_it, options.add_carriage_return);
-        #endif // SEQAN3_WORKAROUND_VIEW_PERFORMANCE
+                written_characters += options.fasta_letters_per_line;
+            }
```
Ok, I know I said we don't care about non-sized or non-random-access ranges. BUT :rofl: 

views::drop is has linear construction on non-random-access ranges in regard to its parameter. And you are going over the whole sequence again every iteration, first the first 70 chars, next time 140 chars... Although you in fact have the iterator where the sequence starts every time! I propose the following loop which is almost identical:

```cpp
            while (it != v.end())
            {
                 auto e = it;
                 std::ranges::advance(e, options.fasta_letters_per_line, std::ranges::end(v));
                 auto r = std::ranges::subrange{it, e};
                 it = stream_it.write_range(r);
            }
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,124 @@ class fast_istreambuf_iterator
+        auto written_chars = copy_res.out - stream_buf->pptr();
+        stream_buf->pbump(written_chars);
+
+        auto remaining_range = rng | views::drop(written_chars);
```
```suggestion
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,124 @@ class fast_istreambuf_iterator
+
+        auto remaining_range = rng | views::drop(written_chars);
+
+        if (remaining_range.begin() == remaining_range.end()) // no more characters to write
```
```suggestion
        if (copy_res.in == std::ranges::end(rng)) // no more characters to write
```

smehringer replied:
This might not work because the input to `std::ranges::copy` is `rng | take(..)` so the iretors are not comparable :(

h-2 replied:
Ah, correct! But You can do :
```cpp
auto tview = rng | views::take(buffer_space);
auto copy_res = std::ranges::copy(tview, stream_buf->pptr());
// ...
if (copy_res.in == std::ranges::end(tview)) // no more characters to write
```

Or, not?

h-2 replied:
Oh, you are right of course! However, this means that you are also not returning an iterator of the original range when calling `return copy_res.in;` and outside this function this will be dangling (it refers to the take_view) and you are assigning it back to the original iterator :boom: (This all just works because our iterators inherit from each other and casting actually does the correct thing by chance)

I would recommend just doing what I proposed on the outside, as well: 

```cpp
                 auto it = std::ranges::begin(rng);
                 auto e = it;
                 std::ranges::advance(e, buffer_space, std::ranges::end(v));
                 auto copy_res = std::ranges::copy(it, e, stream_buf->pptr());
              
                 // ...

                 if (copy_res.in == std::ranges::end(rng)) // no more characters to write
```

Now `copy_res.in` is guaranteed to be of the original iterator type. These changes should have zero overhead for sized random access ranges (advance is in O(1) for them).

h-2 replied:
However, this functions should also be documented that if an rvalue of a range is passed to it and that range is not a safe_range than any returned iterator will still be dangling.

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,124 @@ class fast_istreambuf_iterator
+            throw std::ios_base::failure{"Cannot write to output stream (reached traits::eof() condition)."};
+
+        // drop 1 character that has been written in overflow() and write remaining range recursively
+        return write_range(remaining_range | views::drop(1));
```
```suggestion
        return write_range(remaining_range | views::drop(written_chars + 1));
```

The previous small changes prevent creating two drop-views in case no flattening happens.

h-2 commented on test/performance/io/lowlevel_stream_benchmark.cpp:
```diff
@@ -98,11 +99,97 @@ void read_all(benchmark::State & state)
+
+        it.write_range(cont_rando);
+    }
+}
```
why not integrate this with the rest? Just add an extra tag `tag::seqan3_streambuf_range`!

smehringer replied:
Because it is not a direct comparison to std and seqan2 and also not available for the input streambuf iterator?

h-2 commented on include/seqan3/io/sequence_file/format_embl.hpp:
```diff
@@ -245,34 +255,51 @@ class format_embl
+                    detail::write_eol(stream_it, options.add_carriage_return);
+                }
+
+                it = stream_it.write_range(char_sequence | views::drop(written_characters) | views::take(10));
```
Is this the same problem as below? Non-random-access ranges are parsed by increasing amount of written_characters ever iteration?

smehringer replied:
I wanted to make a card for embl that the format should be improved for non-random access ranges.. because I thought fasta is the more critical one. I can delete the whole embl transformation altogether if this takes too much time.

h-2 replied:
> I wanted to make a card for embl that the format should be improved for non-random access ranges.. 

That's ok, too of course!

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,31 @@ class format_fasta
-                    detail::write_eol(stream_it, options.add_carriage_return);
-            }
-            if (count % options.fasta_letters_per_line != 0)
+                auto end_it = std::ranges::next(it, options.fasta_letters_per_line, std::ranges::end(char_sequence));
```
Here at least, you only need to move by `options.fasta_letters_per_line`, above the amount increases on every iteration.

smehringer replied:
Well yes, there is no way I can circumcent this, except if I write char by char or if I introduce a `write_n(it, n, sentinal)` function to the iterator.. I thought you decided against both?

h-2 replied:
This was just explanation for the previous comment, this if fine.

> I introduce a write_n(it, n, sentinal) function to the iterator.. I thought you decided against both?

As I said before, the `write_n` function won't help us. It makes this potentially faster but the case where the size is not known potentially slower (because of requiring `distance`).

h-2 replied:
[except that this should be replaces with the `couning_next` that I proposed.

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,31 @@ class format_fasta
-            if (count % options.fasta_letters_per_line != 0)
+                auto end_it = std::ranges::next(it, options.fasta_letters_per_line, std::ranges::end(char_sequence));
+                it = stream_it.write_range(std::ranges::subrange<decltype(it), decltype(end_it)>{it, end_it});
                 detail::write_eol(stream_it, options.add_carriage_return);
```
Can you add a note inside the code or as part of the functions documentation that 
a) The performance might be suboptimal for non-random-access-ranges 
b) `views::interleave` is probably the way to go but that needs performance-tuning itself

h-2 replied:
I still think adding a note somewhere about views::interleave wouldn't be too bad. Since that was created exactly for this purpose and actually preserves size and random-access...

smehringer replied:
I did add this comment :) but further up before the whole while loop

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+    /*!\brief Writes a range to the associated output.
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     */
```
Can you document the design decisions and the general mechanics of this function "for future generations"? :wink:

h-2 replied:
[especially return value...]

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+
+        if (rng_it == rng_end) // no more characters to write
+        {
+            if constexpr (std::ranges::viewable_range<range_type>)
```
```suggestion
            if constexpr (std::ranges::safe_range<range_type>)
```

smehringer replied:
`error: ‘safe_range’ is not a member of ‘std::ranges’` ...

h-2 replied:
Hm, it's in range-v3-master but not yet 0.10.0. Then using `seqan3::forwarding_range` for now.

h-2 replied:
Already fixed upstream!

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+
+        auto rng_it = std::ranges::begin(rng);
+        auto rng_end = std::ranges::end(rng);
+
```
```suggestion
size_t size_left = 0;

```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+            auto characters_to_write = std::min<size_t>(std::ranges::size(rng), buffer_space);
+            auto copy_res = std::ranges::copy_n(rng_it, characters_to_write, stream_buf->pptr());
+            rng_it = copy_res.in;
+            stream_buf->pbump(characters_to_write);
```
```suggestion
            stream_buf->pbump(characters_to_write);
            size_left -= characters_to_write;
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,145 @@ class fast_istreambuf_iterator
+        ++rng_it; // drop 1 character that has been written in overflow()
+
+        // write remaining range recursively
+        return write_range(std::ranges::subrange<decltype(rng_it), decltype(rng_end)>{rng_it, rng_end});
```
```suggestion
        if constexpr (std::ranges::sized_range<range_type> &&
                      !std::sized_sentinel_for<std::ranges::sentinel_t<range_type>, std::ranges::iterator_t<range_type>>)
            return write_range(std::ranges::subrange<decltype(rng_it), decltype(rng_end), size_t>{rng_it, rng_end, size_left - 1});
        else
            return write_range(std::ranges::subrange<decltype(rng_it), decltype(rng_end)>{rng_it, rng_end});
```

smehringer replied:
How can a range be sized but the sentinal type not model `sized_sentinal_for` for the iterator?

smehringer replied:
Oh for view take.. :sleepy:

smehringer commented:
I've got the following problem with the sized subrange issue that might be a bug


The following snippet does not compile :( 

```cpp
#include <seqan3/core/debug_stream.hpp>
#include <seqan3/range/views/repeat.hpp>
#include <seqan3/std/ranges>

#include <seqan3/range/views/take.hpp>

int main()
{
    auto v = seqan3::views::repeat('A');
    auto t = v | seqan3::views::take(10);
    auto sr = std::ranges::subrange<decltype(t.begin()), decltype(t.end())>{t.begin(), t.end(), 10};
    seqan3::debug_stream << t;
}

```

The first ctor is rejected because it and sen do not model `sized_sentinel_for` (which is expected)
But the next ctor complains that the required `store_size_` evaluated to false but `store_size_` only checks `K == subrange_kind::sized && !(bool)sized_sentinel_for<S, I>`....
:woman_shrugging: 

```
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/view/subrange.hpp:268:19: note: candidate: ‘constexpr ranges::subrange<I, S, K>::subrange(I2&&, S, ranges::subrange<I, S, K>::size_type) requires  convertible_to_not_slicing_<I2, I> and sized_sentinel_for<S, I> [with I2 = seqan3::detail::view_take<seqan3::detail::repeat_view<char>, false, false>::iterator_type<seqan3::detail::repeat_view<char> >; I = seqan3::detail::view_take<seqan3::detail::repeat_view<char>, false, false>::iterator_type<seqan3::detail::repeat_view<char> >; S = ranges::default_sentinel_t; ranges::subrange_kind K = ranges::subrange_kind::unsized; ranges::subrange<I, S, K>::size_type = long unsigned int]’
  268 |         constexpr CPP_ctor(subrange)(I2 && i, S s, size_type n)( //
      |                   ^~~~~~~~
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/view/subrange.hpp:268:19: note:   constraints not satisfied
In file included from /srv/public/smehringer/seqan3/include/seqan3/std/iterator:25,
                 from /srv/public/smehringer/seqan3/include/seqan3/core/type_traits/iterator.hpp:20,
                 from /srv/public/smehringer/seqan3/include/seqan3/core/type_traits/range.hpp:20,
                 from /srv/public/smehringer/seqan3/include/seqan3/core/detail/debug_stream_range.hpp:18,
                 from /srv/public/smehringer/seqan3/include/seqan3/core/debug_stream.hpp:17,
                 from /srv/public/smehringer/seqan3/test/snippet/range/views/repeat.cpp:1:
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/iterator/concepts.hpp:237:22: note: within ‘template<class S, class I> concept const bool ranges::sized_sentinel_for<S, I> [with S = ranges::default_sentinel_t; I = seqan3::detail::view_take<seqan3::detail::repeat_view<char>, false, false>::iterator_type<seqan3::detail::repeat_view<char> >]’
  237 |     CPP_concept_bool sized_sentinel_for =
      |                      ^~~~~~~~~~~~~~~~~~
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/iterator/concepts.hpp:237:22: note:     with ‘const ranges::default_sentinel_t& s’
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/iterator/concepts.hpp:237:22: note:     with ‘const seqan3::detail::view_take<seqan3::detail::repeat_view<char>, false, false>::iterator_type<seqan3::detail::repeat_view<char> >& i’
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/iterator/concepts.hpp:237:22: note: the required expression ‘((((s - i), (i - s)), requires_<same_as<ranges::iter_difference_t<decltype (i)>, decltype ((s - i))> >), requires_<same_as<ranges::iter_difference_t<decltype (i)>, decltype ((i - s))> >)’ would be ill-formed
In file included from /srv/public/smehringer/seqan3/include/seqan3/std/concepts:25,
                 from /srv/public/smehringer/seqan3/include/seqan3/alphabet/exception.hpp:19,
                 from /srv/public/smehringer/seqan3/include/seqan3/alphabet/concept.hpp:15,
                 from /srv/public/smehringer/seqan3/include/seqan3/core/detail/debug_stream_alphabet.hpp:15,
                 from /srv/public/smehringer/seqan3/include/seqan3/core/debug_stream.hpp:15,
                 from /srv/public/smehringer/seqan3/test/snippet/range/views/repeat.cpp:1:
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/view/subrange.hpp:255:19: note: candidate: ‘constexpr ranges::subrange<I, S, K>::subrange(I2&&, S, ranges::subrange<I, S, K>::size_type) requires  convertible_to_not_slicing_<I2, I> and (store_size_<K, S, I>)() [with I2 = seqan3::detail::view_take<seqan3::detail::repeat_view<char>, false, false>::iterator_type<seqan3::detail::repeat_view<char> >; I = seqan3::detail::view_take<seqan3::detail::repeat_view<char>, false, false>::iterator_type<seqan3::detail::repeat_view<char> >; S = ranges::default_sentinel_t; ranges::subrange_kind K = ranges::subrange_kind::unsized; ranges::subrange<I, S, K>::size_type = long unsigned int]’
  255 |         constexpr CPP_ctor(subrange)(I2 && i, S s, size_type n)( //
      |                   ^~~~~~~~
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/view/subrange.hpp:255:19: note:   constraints not satisfied
/srv/public/smehringer/seqan3/submodules/range-v3/include/range/v3/view/subrange.hpp:255:19: note: ‘(store_size_<K, S, I>)()’ evaluated to false
```

h-2 commented:
> But the next ctor complains that the required store_size_ evaluated to false but store_size_ only checks K == subrange_kind::sized && !(bool)sized_sentinel_for<S, I>....

Well, that's a very clear indication that you should set the third template parameter to `subrange_kind::sized`, or not? :laughing: 

I thought the third parameter was the size_type, but this mechanic is better. I have also updated the suggested code snippet above.

smehringer commented:
> Well, that's a very clear indication that you should set the third template parameter to subrange_kind::sized, or not? :laughing:

Of course I tried that too :roll_eyes: same result

h-2 commented:
The problem is fixed by ericniebler/range-v3#1415

Just use `ranges::subrange` for now.

h-2 commented on include/seqan3/io/sequence_file/format_fasta.hpp:
```diff
@@ -329,47 +326,33 @@ class format_fasta
-            }
-            if (count % options.fasta_letters_per_line != 0)
+                using subrange_t = ranges::subrange<decltype(it), decltype(it), ranges::subrange_kind::sized>;
+                /* Note: Calling detail::counting_next is suboptimal for non-random-access ranges.*/
```
I think the current design is only suboptimal for sized non-random-access ranges. For all non-sized non-random-access ranges parsing outside eliminates the parsing inside (because the subrange being passed in is always sized now). So the solution is really good for 98% of the cases, I think.

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+     * \tparam range_type The type of range to write; Must model std::ranges::forward_range.
+     * \param[in] rng The range to write.
+     * \returns If `range_type` models `std::ranges::forwarding_range` returns an iterator pointing to end of the range
+     *          (rng) else returns void.
```
```suggestion
     *          (rng) else returns `void`.
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+    template <std::ranges::forward_range range_type>
+    void write_range(range_type && rng)
+    {
+        write_range(rng);
```
```suggestion
        write_range(rng); // return value is ignored because iterator would be dangling
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+        it_t rng_it = std::ranges::begin(rng);
+        sen_t rng_end = std::ranges::end(rng);
+
+        size_t buffer_space = stream_buf->epptr() - stream_buf->pptr();
```
```suggestion
        size_t const buffer_space = stream_buf->epptr() - stream_buf->pptr();
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+
+        if constexpr (std::ranges::sized_range<range_type>)
+        {
+            auto characters_to_write = std::min<size_t>(std::ranges::size(rng), buffer_space);
```
```suggestion
            size_t const characters_to_write = std::min<size_t>(std::ranges::size(rng), buffer_space);
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+        // Write remaining range recursively. For sized ranges, ensure that the subrange is also sized.
+        if constexpr (std::ranges::sized_range<range_type> && !std::ranges::sized_sentinel_for<sen_t, it_t>)
+        {
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
```
```suggestion
            // TODO: change to std::ranges::subrange... when updating to range-v3 > 0.10.0
            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
```

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+        if constexpr (std::ranges::sized_range<range_type> && !std::ranges::sized_sentinel_for<sen_t, it_t>)
+        {
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
+            static_assert(forwarding_range<range_type> == forwarding_range<subrange_t>, "not the same");
```
what's dis?

smehringer replied:
ops :D left over debugging

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
+            static_assert(forwarding_range<range_type> == forwarding_range<subrange_t>, "not the same");
+
+            size_t remaining_size = std::ranges::size(rng) - buffer_space - 1;
```
```suggestion
            size_t const remaining_size = std::ranges::size(rng) - buffer_space - 1;
```

h-2 commented on include/seqan3/range/detail/misc.hpp:
```diff
@@ -40,4 +40,34 @@ template <std::ranges::forward_range rng_t>
+ * \param[in] sen The sentinel used as a bound for incrementing `it`.
+ * \returns a std::pair over the incremented iterator and how often `it` was incremented.
+ */
+template <std::ranges::forward_iterator iterator_t, arithmetic size_type, typename sentinal_t>
```
```suggestion
template <std::ranges::forward_iterator iterator_t, std::integral size_type, typename sentinal_t>
```

You cannot increment by `0.4` I think :wink:

h-2 commented on include/seqan3/range/detail/misc.hpp:
```diff
@@ -40,4 +40,34 @@ template <std::ranges::forward_range rng_t>
+//!\cond
+    requires std::ranges::sentinel_for<sentinal_t, iterator_t>
+//!\endcond
+std::pair<iterator_t, size_type> counting_next(iterator_t it, size_type s, sentinal_t sen)
```
```suggestion
std::pair<iterator_t, size_type> counting_next(iterator_t it, size_type const s, sentinal_t sen)
```

:nail_care:

h-2 commented on test/performance/io/CMakeLists.txt:
```diff
@@ -1,5 +1,5 @@
 seqan3_benchmark(format_fasta_benchmark.cpp)
 seqan3_benchmark(format_vienna_benchmark.cpp)
-seqan3_benchmark(lowlevel_stream_input_benchmark.cpp)
+seqan3_benchmark(lowlevel_stream_benchmark.cpp)
```
I would prefer having seperate tests for input and output, because shorter benchmarks means you get to see the desired results quicker!

Input and output are separate for a lot of other things, too.

smehringer replied:
Fine..

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+        ++rng_it; // drop 1 character that has been written in overflow()
+
+        // Write remaining range recursively. For sized ranges, ensure that the subrange is also sized.
+        if constexpr (std::ranges::sized_range<range_type> && !std::ranges::sized_sentinel_for<sen_t, it_t>)
```
```suggestion
        if constexpr (std::ranges::sized_range<range_type>)
```

I think this is enough actually, because the subrange ignores the third parameter if the first two can be substracted from each other.

marehr commented on include/seqan3/io/sequence_file/format_embl.hpp:
```diff
@@ -195,6 +193,19 @@ class format_embl
+     * \param[in] num The number to write.
+     */
+    template <typename stream_iterator_type, arithmetic number_type>
+    auto write_number(stream_iterator_type & stream_it, number_type num)
```
missing test cases

smehringer replied:
now done in iterator test since `write_number` is now a member function of the iterator

h-2 commented on include/seqan3/io/sequence_file/format_embl.hpp:
```diff
@@ -195,6 +193,19 @@ class format_embl
+    {
+        auto res = std::to_chars(&arithmetic_buffer[0], &arithmetic_buffer[0] + sizeof(arithmetic_buffer), num);
+        stream_it.write_range(std::ranges::subrange<char *, char *>(&arithmetic_buffer[0], res.ptr));
+    }
```
For historic reasons! Prior to this PR, we did not have our own stream_it type. I agree that now that we have it, `write_number()` and `write_eol` could become members of it. It could happen as part of this PR, but it could also happen separately.

smehringer commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+     */
+    template <std::ranges::forward_range range_type>
+    //!\cond
+        requires forwarding_range<range_type>
```
Do you mean `safe_range`? Does not compile with gcc-7 :(

marehr replied:
Not good :(

smehringer commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+     * may use memcpy if applicable. Otherwise, a simple for loop iterates over the chunk.
+     *
+     * \attention You can only use the return value (end iterator) if your range type models
+     *            `std::ranges::forwarding_range`.
```
I excluded the below overload for non-forwarding ranges from the documentation and adapted this documentation to fit both overloads.. I thought this was less confusing?

smehringer commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+    template <std::ranges::forward_range range_type>
+    void write_range(range_type && rng)
+    {
+        write_range(rng); // return value is ignored because iterator would be dangling
```
I added a test

marehr replied:
Sorry for the noise, but I can't find the test (edit: Found it). And I still think that this construct is not possible.

edit: My guess is an infinite loop :+1:

smehringer replied:
Well 
1. since there is a test you can see that it works?
2. An lvalue is always a safe range so why should this ever cause an infinite loop as `rng` whithin the function scope is now an lvalue

marehr replied:
Boah this is not intuitive at all; can you add a comment that rng is a lvalue and hence the `safe range` overload is taken

smehringer commented on include/seqan3/range/detail/misc.hpp:
```diff
@@ -40,4 +40,34 @@ template <std::ranges::forward_range rng_t>
+    size_type count = 0;
+    if constexpr (std::sized_sentinel_for<sentinal_t, iterator_t>)
+    {
+        count = std::min<size_type>(sen - it, s);
```
thank you for pointing this out. this actually makes `counting_next` obsolete

marehr commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +171,182 @@ class fast_istreambuf_iterator
+        {
+            // Note: Using ranges here because of bug https://github.com/ericniebler/range-v3/pull/1415 (fixed)
+            // TODO: change to std::ranges::subrange, when updating to range-v3 > 0.10.0
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
```
Still open

marehr replied:
Still open

marehr commented on test/unit/io/stream/fast_streambuf_iterator_test.cpp:
```diff
@@ -0,0 +1,170 @@
+        EXPECT_EQ(rng_it, rng.end());
+        EXPECT_EQ(ostr.str(), std::string_view{"test\ntestest"});
+    }
+
```
a test that tests the overflow behaviour is missing

marehr commented on test/unit/io/stream/fast_streambuf_iterator_test.cpp:
```diff
@@ -0,0 +1,170 @@
+
+    it.write_number(num);
+
+    EXPECT_EQ(ostr.str().substr(0, 12), std::string_view{"54389234"});
```
a test that tests if buffer < 300

smehringer commented:
@marehr sad ping :(

marehr commented on include/seqan3/io/sequence_file/format_embl.hpp:
```diff
@@ -245,34 +241,53 @@ class format_embl
+
+private:
+    //!\brief A buffer to convert numbers into characters with std::to_chars.
+    std::array<char, 50> arithmetic_buffer{};
```
unused?

smehringer replied:
good catch

marehr commented on include/seqan3/io/sequence_file/format_embl.hpp:
```diff
@@ -37,9 +35,8 @@
 #include <seqan3/range/views/take_line.hpp>
 #include <seqan3/range/views/take_until.hpp>
-#include <seqan3/range/views/to.hpp>
+#include <seqan3/range/views/chunk.hpp>
```
is this include even used?

what about the other includes? there seems to be a lot redundancy.

smehringer replied:
it is not used anymore

marehr commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +172,213 @@ class fast_istreambuf_iterator
+            // TODO: change to std::ranges::subrange, when updating to range-v3 > 0.10.0
+            using subrange_t = ranges::subrange<it_t, sen_t, ranges::subrange_kind::sized>;
+            size_t const remaining_size = std::ranges::size(rng) - buffer_space - 1;
+            return write_range(subrange_t{rng_it, rng_end, remaining_size});
```
I just saw this, but why do you recurse (i.e. build up a call-stack) instead of a loop?

smehringer replied:
Yes you are right, I changed it into a for loop

smehringer replied:
This also makes the subrange workaround obsolete

marehr commented on test/unit/io/stream/fast_streambuf_iterator_test.cpp:
Please write tests that executes something and see what happens, not just syntax checking.

My guess is an infinite loop :)

smehringer replied:
There is a test case here:
https://github.com/seqan/seqan3/pull/1472/files#diff-39898526460e7bc8f04d2d99e96e67feR109

smehringer replied:
no infinite loop

smehringer replied:
I added an test. no infinite loop

marehr replied:
Where?

smehringer replied:
wait 'til I push

h-2 commented on include/seqan3/io/stream/iterator.hpp:
```diff
@@ -166,6 +172,204 @@ class fast_istreambuf_iterator
+                {
+                    *stream_buf->pptr() = *it;
+                    stream_buf->pbump(1);
+                }
```
```cpp
                size_t i = 0;
                for (; it != end && i < buffer_space; ++it, ++i)
                    *stream_buf->pptr() = *it;

                stream_buf->pbump(i);
```

Only one virtual function call?

smehringer replied:
pbump is not a virtual function call!

but it might be faster nonetheless

smehringer commented:
@marehr I think I have addressed all of your suggestions now.

We need to keep track of `safe_range` which has been renamed to `borrowed_range`, I'll make a card for this in the range module.

If you approve, do not merge this yet because I want to clean the commit history
